examples/passives/string_dtor.py

```diff
@@ -0,0 +1,103 @@
+from herast.herapi import *
+
+
+class StringDtorScheme(Scheme):
+	def __init__(self):
+		"""
+		pattern of a combination of the forms:
+		if ( v16.capacity >= 0x10)
+		{
+			v12 = *(void **)v16.str_or_ptr;
+			if ( (unsigned __int64)(v16.capacity + 1) >= 0x1000 )
+			{
+				
+			}
+			j_j_free(v12);
+		}
+		or
+		if ( v16 >= 0x10)
+		{
+			v12 = *(void **)v15;
+			if ( (unsigned __int64)(v16 + 1) >= 0x1000 )
+			{
+				
+			}
+			j_j_free(v15);
+		}
+		how destruction of a string inline usually happens
+		"""
+		pattern = IfPat(
+			UgePat(
+				OrPat(
+					VarPat(),
+					MemrefPat(
+						VarPat()
+					),
+				),
+				NumPat(0x10)
+			),
+			BlockPat(
+				AsgInsnPat(
+					VarBindPat("free_var"), 
+					OrPat(
+						PtrPat(
+							MemrefPat(
+								VarBindPat("str_as_str")
+							)
+						),
+						PtrPat(
+							VarBindPat("str_as_str")
+						),
+						VarBindPat("reg_str"),
+						skip_casts=True
+					)
+				),
+				IfPat(
+					UgePat(
+						OrPat(
+							AddPat(
+								VarPat(),
+								NumPat(1)
+							),
+							AddPat(
+								MemrefPat(
+									VarPat()
+								),
+								NumPat(1)
+							),
+							skip_casts=True
+						),
+						NumPat(0x1000),
+					),
+				),
+				CallInsnPat(
+					AnyPat(),
+					OrPat(
+						VarBindPat("free_var"),
+						VarBindPat("str_as_str")
+					),
+				)
+			)
+		)
+		name = "string_dtor"
+		super().__init__(name, pattern)
+
+
+	def on_matched_item(self, item, ctx: PatternContext):
+		str_as_str = ctx.get_var("str_as_str")
+		str_as_var = ctx.get_var("str_as_var")
+		var = None
+		if str_as_str is None:
+			if str_as_var is None:
+				return False
+			else:
+				var = str_as_var
+		else:
+			var = str_as_str
+
+
+		new_item = make_call_helper_instr("string_dtor", var)
+		ctx.modify_instr(item, new_item)

```

Comment:
Sometimes it is better to complitely wipe out destructors for C++ classes, makes decompiled code much clearer :) modify_instr currently accepts argument new_item=None for deletion (this api should be separated)

Reply:
yeah, but usually while reversing I identify that a variable is a string because of this pattern, maybe it will be better if we remove this that pattern and replace the variable type to string

Reply:
Detecting this pattern and changing variable type to std::string is possible too. Also names for schemes are for dynamic behavior, e.g. changing scheme from making new helper instruction to a scheme, that removes instruction

Reply:
Currently there is no api for that (maybe I'll try to add it), I am sorry but I didn't quite understand the second sentence of the last comment and how should I change the code to fit the dynamic approach better

Reply:
You can write ctx.modify_instr(item, new_item) to change instruction or ctx.modify_instr(item, None) to delete instruction. This is not fine and should be changed.

There is name property in schemes to find specific schemes in passive matcher https://github.com/Mizari/herast/blob/main/herast/passive_manager.py#L138 you can get needed scheme, and then change its behavior.
It is just a hint, not a code change request :) you could add flag property to StringDtorScheme should_remove=False. if should_remove is True, then ctx.modify_instr(item, None) is used, otherwise ctx.modify_instr(item, new_item) is used.

In future schemes' name property will be moved from scheme to passive manager, since it is its only purpose and it logically belongs there.